File: Fuzzification Tool/FuzzificationLibrary/Fuzzification.cs
using System;
using System.IO;
using Datasets;

namespace FuzzificationLibrary
{
    public abstract class Fuzzification
    {
        protected Fuzzification(DataSets dataToTransform)
        {
            DataToTransform = dataToTransform;
        }

        public DataSets DataToTransform { get; set; }

        /// <summary>
        ///     dimension, interval, dataitem
        /// </summary>
        public double[][][] Results { get; private set; }

        public int[] Intervals { get; set; }

        /// <summary>
        ///     [i] - dimension,
        ///     [j] - centers
        /// </summary>
        public double[][] Centers { get; set; }

        public int[][] ClassesInInterval { get; set; }

        /// <summary>
        ///     Total entropy - []-dimension, [][] - interval
        /// </summary>
        public double[][] TotalEntropy { get; set; }

        public void Initialize()
        {
            if (DataToTransform != null)
            {
                Intervals = new int[DataToTransform.Attributes];
                Centers = new double[DataToTransform.InputAttributes][];
                TotalEntropy = new double[DataToTransform.InputAttributes][];
                ClassesInInterval = new int[DataToTransform.InputAttributes][];

                for (var i = 0; i < DataToTransform.Attributes; i++)
                    if (DataToTransform.LingvisticAttribute[i] != 0)
                        Intervals[i] = DataToTransform.LingvisticAttribute[i];
                    else
                        Intervals[i] = 2;


                Results = new double[DataToTransform.Attributes][][];

                for (var i = 0; i < DataToTransform.Attributes; i++)
                {
                    Results[i] = new double[Intervals[i]][];

                    for (var j = 0; j < Intervals[i]; j++)
                    {
                        Results[i][j] = new double[DataToTransform.DatasetSize];
                        for (var k = 0; k < DataToTransform.DatasetSize; k++)
                            Results[i][j][k] = 0;
                    }
                }
            }
        }


        public virtual void RunFuzzification()
        {
            Initialize();

            for (var i = 0; i < DataToTransform.Attributes; i++)
                if (DataToTransform.LingvisticAttribute[i] != 0)
                {
                    Intervals[i] = DataToTransform.LingvisticAttribute[i];
                    for (var j = 0; j < DataToTransform.DatasetSize; j++)
                        Results[i]
                            [Convert.ToInt32(DataToTransform.Dataset[j][i])]
                            [j] = 1.0;
                }

            for (var i = 0; i < DataToTransform.Attributes; i++)
                if (DataToTransform.LingvisticAttribute[i] == 0)
                    RunFuzzificationInDimension(i);
        }

        public virtual void RunFuzzificationInDimension(int dimension)
        {
            //  Step 1) Set the initial number of intervals I = 2.
            var interval = SetInitialNumberOfIntervals(dimension);
            TotalEntropy[dimension] = new double[100];
            TotalEntropy[dimension][1] = 99999999999;
            var does_entropy_decrease = 0;
            bool condition;
            do
            {
                //Partition of interval to I = I + 1
                ResizeResultToNewInterval(dimension, interval);

                // Step 2) Locate the centers of intervals.
                Centers[dimension] = DeterminationIntervalsLocation(dimension, Intervals[dimension]);

                // Step 3) Assign membership function for each interval.
                MembershipFunctionAssignment(dimension, interval);

                // Step 4) Compute the total fuzzy entropy of all intervals for I and I - 1 intervals.
                Console.WriteLine(
                    "****************************************************************************************");
                TotalEntropy[dimension][interval] = ComputeTotalFuzzyEntropy(dimension);
                Console.WriteLine("Total fuzzy entropy(" + dimension + "," + interval + "): \t" +
                                  TotalEntropy[dimension][interval]);

                // Step 5) Does the total fuzzy entropy decrease?
                // If the total fuzzy entropy of I intervals is less than that of I - 1 intervals, 
                // then partition again(I := I + 1) and go to Step 2; else go to Step 6.
                condition = ConditionForStopingFuzzificationInDimension(dimension,
                    TotalEntropy[dimension][interval], TotalEntropy[dimension][interval - 1]);
                interval++;
            } while (!condition);

            //Step 6) I - 1 is the number of intervals on specified dimension.
            LastStepInFuzzification(dimension, interval);

            Console.WriteLine("DONE");
        }

        public virtual int SetInitialNumberOfIntervals(int dimension)
        {
            return 2;
        }

        public virtual void LastStepInFuzzification(int dimension, int interval)
        {
            interval = interval - 2;
            ResizeResultToNewInterval(dimension, interval);
            Centers[dimension] = DeterminationIntervalsLocation(dimension, interval);
            MembershipFunctionAssignment(dimension, interval);
        }


        /// <summary>
        ///     Does the total fuzzy entropy decrease?
        ///     If the total fuzzy entropy of I intervals is less than that
        ///     of I - 1 intervals. return true else false;
        /// </summary>
        /// <param name="dimension"></param>
        /// <param name="totalEntropyI"></param>
        /// <param name="totalEntropyIPrevious"></param>
        /// <returns></returns>
        protected abstract double ComputeTotalFuzzyEntropy(int dimension);

        protected virtual bool ConditionForStopingFuzzificationInDimension(int dimension, double totalEntropyI,
            double totalEntropyIPrevious)
        {
            return (totalEntropyI > totalEntropyIPrevious) || (totalEntropyI == 0);
        }

        /// <summary>
        ///     [] center of konkretnehoo data bodu
        /// </summary>
        public abstract double[] DeterminationIntervalsLocation(int dimension, int intervals);

        public virtual void MembershipFunctionAssignment(int dimension, int interval)
        {
            var leftIndex = 0;
            var rightIndex = Centers[dimension].Length - 1;
            for (var i = 0; i < DataToTransform.DatasetSize; i++)
            {
                var x = DataToTransform.Dataset[i][dimension];
                //most left membership function
                var c1 = Centers[dimension][leftIndex];
                var c2 = Centers[dimension][leftIndex + 1];
                double membershipValue = 0;
                if (x <= c1)
                    membershipValue = 1; //alternativa
                else
                    membershipValue = Math.Max(0, 1 - Math.Abs(c1 - x)/Math.Abs(c2 - c1));

                Results[dimension][leftIndex][i] = membershipValue;

                //most right membership function

                var c4 = Centers[dimension][rightIndex];
                var c3 = Centers[dimension][rightIndex - 1];
                if (x <= c4)
                    membershipValue = Math.Max(0, 1 - Math.Abs(c4 - x)/Math.Abs(c4 - c3));
                else if (x > c4)
                    membershipValue = 1; //ALTERNATIVA
                Results[dimension][rightIndex][i] = membershipValue;

                //internal intervals
                for (var j = 1; j < Centers[dimension].Length - 1; j++)
                {
                    c2 = Centers[dimension][j - 1];
                    c3 = Centers[dimension][j];
                    c4 = Centers[dimension][j + 1];

                    if (x <= c3)
                        membershipValue = Math.Max(0, 1 - Math.Abs(c3 - x)/Math.Abs(c3 - c2));
                    else
                        membershipValue = Math.Max(0, 1 - Math.Abs(c3 - x)/Math.Abs(c4 - c3));
                    Results[dimension][j][i] = membershipValue;
                }
            }
        }


        public void ResizeResultToNewInterval(int dimension, int interval)
        {
            Intervals[dimension] = interval;

            Array.Clear(Results[dimension], 0, Results[dimension].Length);

            Results[dimension] = new double[Intervals[dimension]][];

            for (var i = 0; i < Intervals[dimension]; i++)
            {
                Results[dimension][i] = new double[DataToTransform.DatasetSize];
                for (var j = 0; j < DataToTransform.DatasetSize; j++)
                    Results[dimension][i][j] = 0.0;
            }
        }

        public virtual void WriteToFile(string filename = "results.txt")
        {
            using (var w = new StreamWriter(filename))
            {
                if (TotalEntropy != null)
                {
                    w.WriteLine("Total Entropy of dimension on intervals");
                    for (var i = 0; i < TotalEntropy.Length; i++)
                    {
                        for (var j = 0; j < TotalEntropy[i].Length; j++)
                            if (Math.Abs(TotalEntropy[i][j]) > 0.0000001)
                                w.WriteLine("Dimension:  \t" + i + ", Interval: \t" + j + ", Total Entropy = \t" +
                                            TotalEntropy[i][j]);
                        w.WriteLine();
                    }
                    w.WriteLine();
                    w.WriteLine();
                }


                if (Intervals != null)
                {
                    for (var i = 0; i < Intervals?.Length; i++)
                    {
                        w.Write(Intervals[i] + "\t");
                        Console.WriteLine(Intervals[i]);
                    }
                    w.WriteLine();
                }


                if (Centers != null)
                {
                    w.WriteLine("Centers");
                    for (var i = 0; i < Centers.Length; i++)
                    {
                        for (var j = 0; j < Centers[i].Length; j++)
                            w.Write(Math.Round(Centers[i][j], 4).ToString("0.0000") + "\t");
                        w.Write("\t\t");
                    }

                    w.WriteLine();
                }

                if (ClassesInInterval != null)
                    for (var i = 0; i < ClassesInInterval.Length; i++)
                        if (ClassesInInterval[i] != null)
                        {
                            w.WriteLine("Dimension " + i);
                            for (var j = 0; j < ClassesInInterval[i].Length; j++)
                                w.WriteLine(j + " = " + ClassesInInterval[i][j]);
                            w.WriteLine();
                        }

                w.WriteLine(
                    "*****************************************RESULTS***************************************************************");

                if (Results != null)
                {
                    double sum = 0;
                    for (var i = 0; i < DataToTransform?.DatasetSize; i++)
                    {
                        for (var j = 0; j < DataToTransform.Attributes; j++)
                        {
                            for (var k = 0; k < Intervals[j]; k++)
                            {
                                sum += Results[j][k][i];
                                w.Write(Math.Round(Results[j][k][i], 4).ToString("0.0000") + "\t");
                            }
                            w.Write("\t");
                            w.Write("sum(" + Math.Round(sum, 4).ToString("0.0000") + ")" + "\t\t");
                            sum = 0;
                        }
                        w.WriteLine();
                    }
                    w.WriteLine();
                }
            }
        }

        public virtual void WriteResultsToFile(string filename = "results.txt")
        {
            using (var w = new StreamWriter(filename))
            {
                if (Results != null)
                {
                    double sum = 0;
                    for (var i = 0; i < DataToTransform?.DatasetSize; i++)
                    {
                        for (var j = 0; j < DataToTransform.Attributes; j++)
                        {
                            for (var k = 0; k < Intervals[j]; k++)
                            {
                                sum += Results[j][k][i];
                                w.Write(Math.Round(Results[j][k][i], 6).ToString("0.000000") + "\t");
                            }
                            w.Write("\t");
                            sum = 0;
                        }
                        w.WriteLine();
                    }
                    w.WriteLine();
                }
            }
        }
    }
}

File: zdrojove kody v c++/Fuzzification.cpp
#include "Fuzzification.h"
#include "DataSets.h"
#include <algorithm>

using namespace Datasets;

namespace FuzzificationLibrary
{

	Fuzzification::Fuzzification(DataSets *dataToTransform)
	{
		setDataToTransform(dataToTransform);
	}

	DataSets *Fuzzification::getDataToTransform() 
	{
		return _dataToTransform;
	}

	void Fuzzification::setDataToTransform(Datasets::DataSets* value)
	{
		_dataToTransform = value;
	}

	std::vector<std::vector<std::vector<double>>> Fuzzification::getResults() 
	{
		return _results;
	}

	void Fuzzification::setResults(std::vector<std::vector<std::vector<double>>>& value)
	{
		_results = value;
	}

	std::vector<int> Fuzzification::getIntervals() const
	{
		return _intervals;
	}

	void Fuzzification::setIntervals(std::vector<int>& value)
	{
		_intervals = value;
	}

	std::vector<std::vector<double>> Fuzzification::getCenters() const
	{
		return _centers;
	}
	int Fuzzification::getCentersSize(int dimension ) const
	{
		return _centers[dimension].size();
	}


	void Fuzzification::setCenters(std::vector<std::vector<double>>& value)
	{
		_centers = value;
	}

	std::vector<std::vector<int>> Fuzzification::getClassesInInterval() const
	{
		return _classesInInterval;
	}

	void Fuzzification::setClassesInInterval(std::vector<std::vector<int>>& value)
	{
		_classesInInterval = value;
	}

	std::vector<std::vector<double>> Fuzzification::getTotalEntropy() const
	{
		return _totalEntropy;
	}

	void Fuzzification::setTotalEntropy(std::vector<std::vector<double>>& value)
	{
		_totalEntropy = value;
	}

	void Fuzzification::Initialize()
	{
		if (getDataToTransform() != nullptr)
		{
			setIntervals(std::vector<int>(getDataToTransform()->getAttributes()));
			setCenters(std::vector<std::vector<double>>(getDataToTransform()->getInputAttributes()));
			setTotalEntropy(std::vector<std::vector<double>>(getDataToTransform()->getInputAttributes()));
			setClassesInInterval(std::vector<std::vector<int>>(getDataToTransform()->getInputAttributes()));
			for (int i = 0; i < getDataToTransform()->getAttributes(); i++)
			{
				if (getDataToTransform()->getLingvisticAttribute()[i] != 0)
				{
					getIntervals()[i] = getDataToTransform()->getLingvisticAttribute()[i];
				}
				else
				{
					getIntervals()[i] = 2;
				}
			}


			setResults(std::vector<std::vector<std::vector<double>>>(getDataToTransform()->getAttributes()));
			for (int i = 0; i < getDataToTransform()->getAttributes(); i++)
			{
				getResults()[i] = std::vector<std::vector<double>>(getIntervals()[i]);

				for (int j = 0; j < getIntervals()[i]; j++)
				{
					getResults()[i][j] = std::vector<double>(getDataToTransform()->getDatasetSize());
					for (int k = 0; k < getDataToTransform()->getDatasetSize(); k++)
					{
						getResults()[i][j][k] = 0;
					}
				}
			}
		}
	}

	void Fuzzification::RunFuzzification()
	{
		Initialize();

		for (int i = 0; i < getDataToTransform()->getAttributes(); i++)
		{
			if (getDataToTransform()->getLingvisticAttribute()[i] != 0)
			{
				getIntervals()[i] = getDataToTransform()->getLingvisticAttribute()[i];
				for (int j = 0; j < getDataToTransform()->getDatasetSize(); j++)
				{
					getResults()[i][static_cast<int>(getDataToTransform()->getDataset()[j][i])][j] = 1.0;
				}
			}
		}

		for (int i = 0; i < getDataToTransform()->getAttributes(); i++)
		{
			if (getDataToTransform()->getLingvisticAttribute()[i] == 0)
			{
				RunFuzzificationInDimension(i);
			}
		}
	}

	void Fuzzification::RunFuzzificationInDimension(int dimension)
	{
		auto interval = SetInitialNumberOfIntervals(dimension);
		getTotalEntropy()[dimension] = std::vector<double>(100);
		getTotalEntropy()[dimension][1] = 99999999999;
		auto does_entropy_decrease = 0;
		bool condition;
		do
		{
			ResizeResultToNewInterval(dimension, interval);
			getCenters()[dimension] = DeterminationIntervalsLocation(dimension, getIntervals()[dimension]);
			MembershipFunctionAssignment(dimension, interval);
		 getTotalEntropy()[dimension][interval] = ComputeTotalFuzzyEntropy(dimension);
	  condition = ConditionForStopingFuzzificationInDimension(dimension, getTotalEntropy()[dimension][interval], getTotalEntropy()[dimension][interval - 1]);
			interval++;
		} while (!condition);
		LastStepInFuzzification(dimension, interval);
	}

	int Fuzzification::SetInitialNumberOfIntervals(int dimension)
	{
		return 2;
	}

	void Fuzzification::LastStepInFuzzification(int dimension, int interval)
	{
		interval = interval - 2;
		ResizeResultToNewInterval(dimension, interval);
		getCenters()[dimension] = DeterminationIntervalsLocation(dimension, interval);
		MembershipFunctionAssignment(dimension, interval);
	}

	bool Fuzzification::ConditionForStopingFuzzificationInDimension(int dimension, double totalEntropyI, double totalEntropyIPrevious)
	{
		return (totalEntropyI > totalEntropyIPrevious) || (totalEntropyI == 0);
	}

	void Fuzzification::MembershipFunctionAssignment(int dimension, int interval)
	{
		auto leftIndex = 0;
		auto rightIndex = getCenters()[dimension].size() - 1;
		for (int i = 0; i < getDataToTransform()->getDatasetSize(); i++)
		{
			auto x = getDataToTransform()->getDataset()[i][dimension];
			//most left membership function
			auto c1 = getCenters()[dimension][leftIndex];
			auto c2 = getCenters()[dimension][leftIndex + 1];
			double membershipValue = 0;
			if (x <= c1)
			{
				membershipValue = 1; //alternativa
			}
			else
			{
				membershipValue = (1 - std::abs(c1 - x) / std::abs(c2 - c1));
				if(membershipValue<0)
				{
					membershipValue = 0; 
				}
			}

			getResults()[dimension][leftIndex][i] = membershipValue;

			//most right membership function

			auto c4 = getCenters()[dimension][rightIndex];
			auto c3 = getCenters()[dimension][rightIndex - 1];
			if (x <= c4)
			{
				membershipValue = ( 1 - std::abs(c4 - x) / std::abs(c4 - c3));
				if (membershipValue<0)
				{
					membershipValue = 0;
				}
			}
			else if (x > c4)
			{
				membershipValue = 1; //ALTERNATIVA
			}
			getResults()[dimension][rightIndex][i] = membershipValue;
			//internal intervals
			for (int j = 1; j < getCentersSize(dimension)-1; j++)
			{
				c2 = getCenters()[dimension][j - 1];
				c3 = getCenters()[dimension][j];
				c4 = getCenters()[dimension][j + 1];

				if (x <= c3)
				{
					membershipValue =  1 - std::abs(c3 - x) / std::abs(c3 - c2);
					if (membershipValue<0)
					{
						membershipValue = 0;
					}
				}
				else
				{
					membershipValue =  1 - std::abs(c3 - x) / std::abs(c4 - c3);
					if (membershipValue<0)
					{
						membershipValue = 0;
					}
				}
				getResults()[dimension][j][i] = membershipValue;
			}
		}
	}

	void Fuzzification::ResizeResultToNewInterval(int dimension, int interval) const
	{
		getIntervals()[dimension] = interval;

		getResults()[dimension].erase(getResults()[dimension].begin());

		getResults()[dimension] = std::vector<std::vector<double>>(getIntervals()[dimension]);

		for (int i = 0; i < getIntervals()[dimension]; i++)
		{
			getResults()[dimension][i] = std::vector<double>(getDataToTransform()->getDatasetSize());
			for (int j = 0; j < getDataToTransform()->getDatasetSize(); j++)
			{
				getResults()[dimension][i][j] = 0.0;
			}
		}
	}

	}
